Point cloud editing: rewrite COPC when committing changes

wonder-sk opened the pull request:
This PR extends work started in #59973 - it adds functionality to store changes to attributes of a point cloud to a COPC file.

There is a new class `QgsCopcUpdate` that:
1. reads and existing COPC file
2. writes a new COPC file with some chunks (nodes) with updated content

Writing of new COPC files is fast, because typically only a small amount of points are modified and so most of the chunks are copied as they were. There are also no changes to the hierarchy.

QgsPointCloudLayer::commitChanges() has been modified to use QgsCopcUpdate. A new file is created next to the original file. The original file is then renamed, and newly created file is renamed to effectively replace it. In the final step, the original file is removed.

wonder-sk committed:
Point cloud editing: write COPC when committing changes

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

wonder-sk committed:
ci fixes

wonder-sk committed:
fix build

uclaros commented on src/core/pointcloud/qgspointcloudlayereditutils.cpp:
```diff
@@ -59,27 +64,171 @@ bool QgsPointCloudLayerEditUtils::changeAttributeValue( const QgsPointCloudNodeI
+
+QByteArray QgsPointCloudLayerEditUtils::updateChunkValues( QgsCopcPointCloudIndex *copcIndex, const QByteArray &chunkData, const QgsPointCloudAttribute &attribute, double newValue, QgsPointCloudNodeId k, QVector<int> pointIndices )
+{
+  // set new classification value for the given points in voxel and return updated chunk data
```
Remove comment, or update and move to header file?

uclaros commented on src/core/pointcloud/qgspointcloudlayereditutils.h:
```diff
@@ -63,6 +65,9 @@ class CORE_EXPORT QgsPointCloudLayerEditUtils
 
   private:
+
+    QByteArray updateChunkValues( QgsCopcPointCloudIndex *copcIndex, const QByteArray &chunkData, const QgsPointCloudAttribute &attribute, double newClassValue, QgsPointCloudNodeId k, QVector<int> pointIndices );
```
```suggestion
    QByteArray updateChunkValues( QgsCopcPointCloudIndex *copcIndex, const QByteArray &chunkData, const QgsPointCloudAttribute &attribute, double newClassValue, QgsPointCloudNodeId k, const QVector<int> &pointIndices );
```

uclaros commented on src/core/pointcloud/qgspointcloudeditingindex.cpp:
```diff
@@ -114,15 +124,63 @@ QgsPointCloudBlockRequest *QgsPointCloudEditingIndex::asyncNodeData( const QgsPo
+  {
+    if ( errorMessage )
+      *errorMessage = "Removal of the old COPC failed!";
+    // TODO: cleanup here as well?
```
Is the comment still relevant?

wonder-sk replied:
yeah, it's unclear what to do... removal of the old (now renamed) file failed, but in fact from the user's point of view the commit succeeded - so I'm not sure what the most appropriate here. Probably it does not matter because this would never happen, and if there's a problem, it would fail earlier :slightly_smiling_face:

uclaros commented on src/core/pointcloud/qgspointcloudeditingindex.cpp:
```diff
@@ -138,5 +196,19 @@ bool QgsPointCloudEditingIndex::updateNodeData( const QHash<QgsPointCloudNodeId,
     mEditedNodeData[it.key()] = it.value();
   }
 
+  const QList<QgsPointCloudNodeId> modifiedNodesIds = data.keys();
+  QSet<QgsPointCloudNodeId> modifiedNodesSet( modifiedNodesIds.constBegin(), modifiedNodesIds.constEnd() );
+
+  // get rid of cached keys that got modified
+  {
+    QMutexLocker locker( &sBlockCacheMutex );
+    const QList<QgsPointCloudCacheKey> cacheKeys = sBlockCache.keys();
+    for ( QgsPointCloudCacheKey cacheKey : cacheKeys )
+    {
+      if ( cacheKey.uri() == mUri && modifiedNodesSet.contains( cacheKey.node() ) )
+        sBlockCache.remove( cacheKey );
+    }
+  }
```
`QHash::keys()` creates a copy of the keys.
Could we use `data.contains( cacheKey.node() )` instead of going though QList and QSet?

Then for the second case of `sBlockCache.keys()` we could use a hash iterator and `QHash::erase()`?

uclaros reviewed (commented):
Nicely done!

Just some tidying up comments.

Also a question: We mostly pass `QgsPointCloudNodeId` by const ref and name it `n`, while here it is mostly passed by val and named `k`. Should we stick with one or the other? With `QgsPointCloudNodeId` being practically 4 ints it doesn't really matter, right?

wonder-sk committed:
review fixes

wonder-sk commented:
Thanks for the review!

> We mostly pass QgsPointCloudNodeId by const ref and name it n, while here it is mostly passed by val and named k.

yeah I have changed it now to const ref for consistency, and also renamed `k` (as for voxel Key) to `n` ...
